File: VineLine/VineLineDrawingView.m
//
//  VineLineDrawingView.m
//  VineLine
//
//  Created by Michael Behan on 28/01/2014.
//  Copyright (c) 2014 Michael Behan. All rights reserved.
//

#import "VineLineDrawingView.h"
#import "VineLine.h"
#import <QuartzCore/QuartzCore.h>

@interface VineLineDrawingView()
{
    NSMutableArray *vineLines;
    UIColor *vineColor;
}
@end

@implementation VineLineDrawingView

-(id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        // Initialization code
        
        vineLines = [[NSMutableArray alloc] init];
        vineColor = [UIColor colorWithRed:78/255.0 green:154/255.0 blue:131/255.0 alpha:1];
        
        UIPanGestureRecognizer *pgr = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(handlePan:)];
        
        [self addGestureRecognizer:pgr];
        
        _leafSize = 10.0;
        _branchSeperation = 100.0;
        _vineWidth = 5.0;
        _branchLength = 50.0;
    }
    return self;
}

-(void)handlePan:(UIPanGestureRecognizer *)gestureRecognizer
{
    if(gestureRecognizer.state == UIGestureRecognizerStateBegan)
    {
        VineLine *newVineLine = [[VineLine alloc] init];
        newVineLine.delegate = self;
        newVineLine.lineWidth = _vineWidth;
        newVineLine.minBranchSeperation = _branchSeperation;
        newVineLine.maxBranchLength = _branchLength;
        newVineLine.leafSize = _leafSize;
        [newVineLine moveToPoint:[gestureRecognizer locationInView:self]];
        
        [vineLines addObject:newVineLine];
    }
    else if(gestureRecognizer.state == UIGestureRecognizerStateChanged)
    {
        VineLine *currentLine = [vineLines lastObject];
        
        [currentLine addLineToPoint:[gestureRecognizer locationInView:self]];
    }
    else
    {
        
    }
    [self setNeedsDisplay];
    
}

-(void)vineLineDidCreateBranch:(VineBranch *)branchPath
{
    CAShapeLayer *branchShape = [CAShapeLayer layer];
    branchShape.path = branchPath.CGPath;
    branchShape.fillColor = [UIColor clearColor].CGColor;
    branchShape.strokeColor = vineColor.CGColor;
    branchShape.lineWidth = branchPath.lineWidth;
    
    [self.layer addSublayer:branchShape];
    
    CABasicAnimation *branchGrowAnimation = [CABasicAnimation animationWithKeyPath:@"strokeEnd"];
    branchGrowAnimation.duration = 1.0;
    branchGrowAnimation.fromValue = [NSNumber numberWithFloat:0.0];
    branchGrowAnimation.toValue = [NSNumber numberWithFloat:1.0];
    [branchShape addAnimation:branchGrowAnimation forKey:@"strokeEnd"];
}


// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect
{
    // Drawing code
    
    [vineColor setStroke];
    
    for(VineLine *vineLine in vineLines)
    {
        [vineLine stroke];
        /*
        for(UIBezierPath *branchLine in vineLine.branchLines)
        {
            [branchLine stroke];
        }
         */
    }
}


@end

File: VineLine/VineLine/VineLineDrawingView.swift
//
//  DrawingView.swift
//  VineLine
//
//  Created by Michael Behan on 17/03/2019.
//  Copyright © 2019 Michael Behan. All rights reserved.
//

import UIKit

class VineLineDrawingView: UIView, VineDelegate {
    func vineDidCreate(branch: Branch) {

        let branchShape = CAShapeLayer()
        branchShape.path = branch.cgPath
        branchShape.fillColor = UIColor.clear.cgColor
        branchShape.strokeColor = branch.color.cgColor
        branchShape.lineWidth = branch.lineWidth

        layer.addSublayer(branchShape)

        let branchGrowAnimation = CABasicAnimation(keyPath: "strokeEnd")
        branchGrowAnimation.duration = 1.0
        branchGrowAnimation.fromValue = 0.0
        branchGrowAnimation.toValue = 1.0
        branchShape.add(branchGrowAnimation, forKey: "strokeEnd")

    }


    private let lineWidth = CGFloat(5.0)
    private var activeStrokes = [UITouch: Vine]()
    private var strokes = [Vine]()

    var leafSize = 10.0
    var branchSeperation = 100.0
    var vineWidth = 5.0
    var branchLength = 50.0
    
    // MARK:- Init

    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }

    func setup(){
        isMultipleTouchEnabled = true
    }
    
    // MARK:- Touch handling
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        self.isMultipleTouchEnabled = true
        for touch in touches {
            let location = touch.location(in: self)
            let newVine = Vine()
            newVine.lineCapStyle = .round
            newVine.lineWidth = lineWidth
            newVine.move(to: location)
            newVine.addLine(to: location)
            newVine.delegate = self
            newVine.lineWidth = CGFloat(vineWidth)
            newVine.minBranchSeperation = branchSeperation
            newVine.maxBranchLength = branchLength
            newVine.leafSize = leafSize
            newVine.color = UIColor.random
            activeStrokes[touch] = newVine
        }
    }

    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        for touch in touches {
            activeStrokes[touch]!.addLine(to: touch.location(in: self))
        }
        self.setNeedsDisplay()
    }

    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        for touch in touches {
            let line = activeStrokes.removeValue(forKey: touch)!
            line.move(to: touch.location(in: self))
            strokes.append(line)
        }
        self.setNeedsDisplay()
    }
    
    // MARK:- Drawing
    
    override func draw(_ rect: CGRect) {
        for line in strokes {
            line.color.setStroke()
            line.stroke()
        }
        
        for line in activeStrokes.values {
            line.color.setStroke()
            line.stroke()
        }
    }
}
